spec/controllers/posts_controller_spec.rb

```diff
@@ -44,59 +44,87 @@
       end
 
       it "returns an error if seller is not eligible to send emails" do
         allow_any_instance_of(User).to receive(:sales_cents_total).and_return(Installment::MINIMUM_SALES_CENTS_VALUE - 1)
 
-        @purchase.create_url_redirect!
-        expect(PostSendgridApi).to_not receive(:process)
-        get :send_for_purchase, params: { id: @post.external_id, purchase_id: @purchase.external_id }
+        post :send_for_purchase, params: { id: @post.external_id, purchase_id: @purchase.external_id }
         expect(response).to have_http_status(:unauthorized)
         expect(response.parsed_body).to eq("message" => "You are not eligible to resend this email.")
       end
 
       it "returns 404 if no purchase" do
-        expect(PostSendgridApi).to_not receive(:process)
-        expect do
-          get :send_for_purchase, params: { id: @post.external_id, purchase_id: "hello" }
-        end.to raise_error(ActiveRecord::RecordNotFound)
+        post :send_for_purchase, params: { id: @post.external_id, purchase_id: "hello"  }
+
+        expect(response).to have_http_status(:not_found)
+        expect(response.parsed_body).to eq("success" => false, "error" => "Not found")
       end
 
-      it "returns success and redelivers the installment" do
-        @purchase.create_url_redirect!
-        expect(PostSendgridApi).to receive(:process).with(
-          post: @post,
-          recipients: [{
-            email: @purchase.email,
-            purchase: @purchase,
-            url_redirect: @purchase.url_redirect,
-          }]
-        )
-        get :send_for_purchase, params: { id: @post.external_id, purchase_id: @purchase.external_id }
-        expect(response).to be_successful
-        expect(response).to have_http_status(:no_content)
+      it "returns success" do
+        allow(SendPostsForPurchaseService).to receive(:send_post)
 
-        # when the purchase part of a subscription
-        membership_purchase = create(:membership_purchase, link: create(:membership_product, user: @post.seller))
-        membership_purchase.create_url_redirect!
-        expect(PostSendgridApi).to receive(:process).with(
-          post: @post,
-          recipients: [{
-            email: membership_purchase.email,
-            purchase: membership_purchase,
-            url_redirect: membership_purchase.url_redirect,
-            subscription: membership_purchase.subscription,
-          }]
-        )
-        get :send_for_purchase, params: { id: @post.external_id, purchase_id: membership_purchase.external_id }
+        post :send_for_purchase, params: { id: @post.external_id, purchase_id: @purchase.external_id }
+        expect(SendPostsForPurchaseService).to have_received(:send_post).with(post: @post, purchase: @purchase)
         expect(response).to be_successful
         expect(response).to have_http_status(:no_content)
       end
     end
+
+    describe "POST send_missed_posts" do
```

Comment:
Similar approach over here. 

Checking for email delivery in service specs and binding with service over here